Source/WebDriver/WebDriverService.cpp

```diff
@@ -475,46 +475,42 @@ void WebDriverService::handleMessage(WebSocketMessageHandler::Message&& message,
         return;
     }
 
+    const auto& asObject = parsedMessageValue->asObject();
+    if (!asObject) {
+        RELEASE_LOG_ERROR(WebDriverBiDi, "WebDriver handle BiDi message: Expected object.");
+        completionHandler(WebSocketMessageHandler::Message::fail(CommandResult::ErrorCode::InvalidArgument, connection));
+        return;
+    }
+
+    std::optional<int> commandId = asObject->getInteger("id"_s);
+    if (!commandId) {
+        RELEASE_LOG_ERROR(WebDriverBiDi, "Missing command ID.");
+        completionHandler(WebSocketMessageHandler::Message::fail(CommandResult::ErrorCode::InvalidArgument, connection, "Missing command id"_s));
+        return;
+    }
+
     BidiCommandHandler handler;
-    unsigned id = 0;
     RefPtr<JSON::Object> parameters;
-    if (!findBidiCommand(parsedMessageValue, &handler, id, parameters)) {
-        RELEASE_LOG(WebDriverBiDi, "Failed to find appropriate BiDi command");
-        std::optional<int> commandId;
-        if (auto parsedMessageObject = parsedMessageValue->asObject()) {
-            auto parsedCommandId = parsedMessageObject->getInteger("id"_s);
-            if (parsedCommandId && *parsedCommandId >= 0)
-                commandId = parsedCommandId;
-        }
-
-        auto errorCode = CommandResult::ErrorCode::UnknownCommand;
-        auto errorReply = WebSocketMessageHandler::Message::fail(errorCode, connection, { "Command not supported"_s }, commandId);
-        completionHandler(WTFMove(errorReply));
+    if (!findBidiCommand(asObject, &handler, parameters)) {
+        RELEASE_LOG(WebDriverBiDi, "Failed to find appropriate BiDi command on WebDriver service. Relaying to the browser.");
+        m_session->relayBidiCommand(makeString(message.payload), *commandId, [completionHandler = WTFMove(completionHandler), message](WebSocketMessageHandler::Message&& resultMessage) {
+            resultMessage.connection = message.connection;
```

Comment:
It seems a little weird to me that we keep the client context by setting the connection on each `resultMessage`. If `connection` was wrapping some networking library objects, I would be concerned about leaking those objects if we accidentally retained or copied resultMessage. What we really want to know is which *websocket* client originated the command. I would prefer that the WebSocketServer vend an opaque `ClientIdentifier` to represent WebSocket clients instead of connection objects.

Reply:
Good point. I'll modify to save just the `sessionID` associated to the request in the closure instead, given we can get the associated connection inside the callback.

As for `clientID`, we can add a proper scheme when adding static methods, as I mentioned above. WDYT?